### ISSUE TITLE
Enhancement: Documentation for states

### ISSUE THREAD

**Issue by amyspark-ng** [FunkinCrew/Funkin#3334]
### Issue Checklist

- [X] I have properly named the enhancement
- [X] I have checked the issues/discussions pages to see if the enhancement has been previously suggested

### What is your suggestion, and why should it be implemented?

Was wondering if documentation for states is something we're getting sooner or later, it would be nice to have a look at some variables that are for example in playstate

**Comment by amyspark-ng** [comment]
Like being able to look at the documentation and see which events we can work with, like onNoteHit, i didn't know needsReset was a thing! to make mods you mostly need to look at the code rather than the documentation in my opinion

**Comment by EliteMasterEric** [comment]
I definitely want to work on documenting this a bit more, I'm experimenting with automatic documentation which should speed things along

**Comment by amyspark-ng** [comment]
That's great! glad to hear that

the other day I wanted to make a small mod that makes it so if you press Ctrl + R the song resets

managed this by finding the needsReset variable in playstate, also wanted to make it so it closes pause screen and figured out by finding the closeSubstate function, seems like all of them are on files that are not exactly documented

also was wondering if it's possible to get some kind of intellisense while working on mods like this? so i can see all the variables that exist on all the classes

**Comment by EliteMasterEric** [comment]
> also was wondering if it's possible to get some kind of intellisense while working on mods like this? so i can see all the variables that exist on all the classes

Haxe is supposed to provide Intellisense, but various issues have made it unstable in the Funkin' project for a long time. It's something we're looking into.

**Comment by amyspark-ng** [comment]
alright thank you! I guess I'll leave it open until it happens

**Comment by Lasercar** [comment]
> > also was wondering if it's possible to get some kind of intellisense while working on mods like this? so i can see all the variables that exist on all the classes
> 
> Haxe is supposed to provide Intellisense, but various issues have made it unstable in the Funkin' project for a long time. It's something we're looking into.

Oh, so that's why it wasn't working. I thought it was an issue on my end. It didn't help that I didn't know that I could enter a class name in the top search bar to go directly to it so I could then find the function - I was instead searching for the function name in the search tab...

**Comment by NotHyper-474** [comment]
Yep, I personally though it was an issue with vshaxe and even made an issue there. Although I found that for whatever reason opening `.haxelib/flixel/git/flixel/group/FlxSpriteGroup.hx` then removing `:@generic` from line 633 seems to solve the issue for me, albeit IntelliSense will still be quite slow.
